src/Plugin/resource/DataProvider/DataProviderEntity.php

```diff
@@ -592,28 +598,73 @@ protected function queryForListSort(\EntityFieldQuery $query) {
+          // the target of that is the current resource.
+          continue;
+        }
+        $this->addNestedFilter($filter, $query);
+        continue;
       }
       if (!$property_name = $resource_field->getProperty()) {
         throw new BadRequestException(sprintf('The current filter "%s" selection does not map to any entity property or Field API field.', $filter['public_field']));
       }
+
       if (field_info_field($property_name)) {
         if (in_array(strtoupper($filter['operator'][0]), array('IN', 'BETWEEN'))) {
-          $query->fieldCondition($property_name, $resource_field->getColumn(), $filter['value'], $filter['operator'][0]);
+          $query->fieldCondition($property_name, $resource_field->getColumn(), $this->getReferencedIds($filter['value'], $resource_field), $filter['operator'][0]);
           continue;
         }
         for ($index = 0; $index < count($filter['value']); $index++) {
-          $query->fieldCondition($property_name, $resource_field->getColumn(), $filter['value'][$index], $filter['operator'][$index]);
+          // If referencing an entity by an alternate ID, change the user
+          // provided value by the entity value.
+          $query->fieldCondition($property_name, $resource_field->getColumn(), $this->getReferencedId($filter['value'][$index], $resource_field), $filter['operator'][$index]);
         }
       }
       else {
         $column = $this->getColumnFromProperty($property_name);
+        if (in_array(strtoupper($filter['operator'][0]), array('IN', 'BETWEEN'))) {
+          $query->propertyCondition($column, $this->getReferencedIds($filter['value'], $resource_field), $filter['operator'][0]);
+          continue;
+        }
         for ($index = 0; $index < count($filter['value']); $index++) {
-          $query->propertyCondition($column, $filter['value'][$index], $filter['operator'][$index]);
+          $query->propertyCondition($column, $this->getReferencedId($filter['value'][$index], $resource_field), $filter['operator'][$index]);
+        }
+      }
+    }
+  }
+
+  /**
+   * Validates the query parameters.
+   *
+   * @param array $filters
+   *   The parsed filters.
+   *
+   * @throws BadRequestException
+   *   When there is an invalid target for relational filters.
+   */
+  protected function validateFilters(array $filters) {
+    foreach ($filters as $filter) {
+      if (empty($filter['target'])) {
+        continue;
+      }
+      // If the target is not a part of the field, then raise an error.
+      $field_name_parts = explode('.', $filter['public_field']);
+      $target_parts = explode('.', $filter['target']);
+      foreach ($target_parts as $delta => $target_part) {
+        if ($target_part != $field_name_parts[$delta]) {
+          // There is a discrepancy between target and field name.
+          throw new BadRequestException(sprintf('The target "%s" should be a part of the field name "%s".', $filter['target'], $filter['public_field']));
```

Comment:
Actually I like `target`, more than `sub-field` since you are specifying what resource you are filtering.

See http://gist.github.com/mateu-aguilo-bosch/1acbced683908ff631e3.